subsys/bluetooth/mesh/rpl.c

```diff
@@ -164,183 +113,30 @@ void bt_mesh_rpl_clear(void)
 {
 	BT_DBG("");
 
-	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
-		schedule_rpl_clear();
-	} else {
-		(void)memset(replay_list, 0, sizeof(replay_list));
-	}
-}
-
-static struct bt_mesh_rpl *bt_mesh_rpl_find(uint16_t src)
-{
-	int i;
-
-	for (i = 0; i < ARRAY_SIZE(replay_list); i++) {
-		if (replay_list[i].src == src) {
-			return &replay_list[i];
-		}
-	}
-
-	return NULL;
-}
-
-static struct bt_mesh_rpl *bt_mesh_rpl_alloc(uint16_t src)
-{
-	int i;
-
-	for (i = 0; i < ARRAY_SIZE(replay_list); i++) {
-		if (!replay_list[i].src) {
-			replay_list[i].src = src;
-			return &replay_list[i];
-		}
-	}
-
-	return NULL;
+	bt_mesh_rpl_strg_clear_all();
 }
 
 void bt_mesh_rpl_reset(void)
 {
-	int i;
+	struct bt_mesh_rpl *rpl = NULL;
 
 	/* Discard "old old" IV Index entries from RPL and flag
 	 * any other ones (which are valid) as old.
 	 */
-	for (i = 0; i < ARRAY_SIZE(replay_list); i++) {
-		struct bt_mesh_rpl *rpl = &replay_list[i];
+	for (int i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
+		rpl = bt_mesh_rpl_strg_iter(rpl);
```

Comment:
If you are going to encapsulate RPL iteration, then it should be done properly, so that the caller doesn't need to know details like `CONFIG_BT_MESH_CRPL`, i.e. this could instead look something like:
```c
for (struct bt_mesh_rpl *rpl = bt_mesh_strg_iter(NULL); rpl != NULL; rpl = bt_mesh_strg_iter(rpl)) {
	...;
}
```
That said, I think you might get a bit more efficient code (and also be more consistent with the iterators we have elsewhere within the Zephyr Bluetooth stack) by introducing a foreach function with a callback pointer:

```c
/* Callback can return false to abort iteration, or true to continue iterating */
void bt_mesh_rpl_foreach(bool (*func)(struct bt_mesh_rpl *rpl, void *user_data), void *user_data);
```

Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the `replay_list` array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

If you're only intending to encapsulate the persistent storage, it would seem more appropriate for `rpl.c` to retain ownership of the runtime storage. If you need to do iteration in the storage backend, then the backend needs to call the iterator function (which would be implemented in `rpl.c`).

If you're intending to encapsulate both runtime and persistent storage, it looks like the purpose of `rpl.c` becomes really small (just a bit over 100 lines of code), and I wonder if it'd make more sense to just push all responsibility to the storage backend.

Note: I'm just thinking out loud here - don't take these as some final commandments, but rather let me know your thoughts!

Reply:
> Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the replay_list array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

Yes, this is the idea. A storage mechanism that we are planning to use in nRF Connect SDK for RPL will replace both, the runtime and persistence management. At the same time we want to keep the internal API implementation unchanged as much as possible to avoid potential bugs. Do you think this implementation will impact on performance?

Reply:
In that case it feels a bit like we might not need the additional layer of abstraction, rather the existing `rpl.c` could be replaced with one of several possible `rpl_<backend>.c` implementations. If there's substantial amounts of code that multiple backends would need to duplicate, we could think of some alternate solutions to that, e.g. inline functions in `rpl.h` or similar.

Reply:
> Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the replay_list array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

> If you're only intending to encapsulate the persistent storage, it would seem more appropriate for rpl.c to retain ownership of the runtime storage. If you need to do iteration in the storage backend, then the backend needs to call the iterator function (which would be implemented in rpl.c).

> If you're intending to encapsulate both runtime and persistent storage, it looks like the purpose of rpl.c becomes really small (just a bit over 100 lines of code), and I wonder if it'd make more sense to just push all responsibility to the storage backend.

We have been working on an alternate persistent storage implementation in our Downstream. The storage mechanism for this module will write the entire RPL to flash in a single operation at shutdown, meaning that the module needs access to the memory object.

We have been back and forth on this in an internal review, and here are the main points:

If `rpl.c` owns the memory object for the RPL, the storage module would either need API access the memobj, or duplicate the entire RPL before storing it. To provide direct access to the memobj has previously been strongly discouraged since it will provide direct access to senitive data ([Link](https://github.com/zephyrproject-rtos/zephyr/pull/43441#discussion_r820671571)). Duplicating the list is neither a good option since it will double the amount of memory needed to ensure that the data is stored, especially when the RPL is large.

However, if the storage module owns the memobj then all these issues should disapear.

Reply:
> In that case it feels a bit like we might not need the additional layer of abstraction, rather the existing `rpl.c` could be replaced with one of several possible `rpl_<backend>.c` implementations. If there's substantial amounts of code that multiple backends would need to duplicate, we could think of some alternate solutions to that, e.g. inline functions in `rpl.h` or similar.

Ok, on first thought I'm not necessarily opposed to this. It would remove the need to provide iteration API over the RPL array all together.

Reply:
How about adding few functions to replace `bt_mesh_rpl_strg_iter`:
- _find first entry with such address or first empty entry_ for `bt_mesh_rpl_check`;
- _get next non-empty entry or return NULL_ for `bt_mesh_rpl_reset`;

Reply:
@PavelVPV that doesn’t sound like a bad approach. Would be good to see the code first however.